src/System.Threading.Channels/src/System/Threading/Channels/AsyncOperation.cs

```diff
@@ -0,0 +1,370 @@
+            }
+
+            return false;
+        }
+
+        public void OnCompleted(Action<object> continuation, object state, short token, ValueTaskSourceOnCompletedFlags flags)
+        {
+            if (_currentId != token)
+            {
+                ThrowIncorrectCurrentIdException();
+            }
+
+            // We need to store the state before the CompareExchange, so that if it completes immediately
+            // after the CompareExchange, it'll find the state already stored.  If someone misuses this
+            // and schedules multiple continuations erroneously, we could end up using the wrong state.
+            // Make a best-effort attempt to catch such misuse.
+            if (_continuationState != null)
+            {
+                ThrowMultipleContinuations();
+            }
+            _continuationState = state;
+
+            Debug.Assert(_executionContext == null);
+            if ((flags & ValueTaskSourceOnCompletedFlags.FlowExecutionContext) != 0)
+            {
+                _executionContext = ExecutionContext.Capture();
+            }
+
+            Debug.Assert(_schedulingContext == null);
+            SynchronizationContext sc = null;
+            TaskScheduler ts = null;
+            if ((flags & ValueTaskSourceOnCompletedFlags.UseSchedulingContext) != 0)
+            {
+                sc = SynchronizationContext.Current;
+                if (sc != null && sc.GetType() != typeof(SynchronizationContext))
+                {
+                    _schedulingContext = sc;
+                }
+                else
+                {
+                    ts = TaskScheduler.Current;
+                    if (ts != TaskScheduler.Default)
+                    {
+                        _schedulingContext = ts;
+                    }
+                }
+            }
+
+            Action<object> prevContinuation = Interlocked.CompareExchange(ref _continuation, continuation, null);
+            if (prevContinuation != null)
+            {
+                if (prevContinuation != s_completedSentinel)
+                {
+                    ThrowMultipleContinuations();
+                }
+
+                Debug.Assert(IsCompleted, $"Expected IsCompleted, got {(States)_state}");
+                if (sc != null)
+                {
+                    sc.Post(s =>
+                    {
+                        var t = (Tuple<Action<object>, object>)s;
+                        t.Item1(t.Item2);
+                    }, Tuple.Create(continuation, state));
+                }
+                else
+                {
+                    Task.Factory.StartNew(continuation, state, CancellationToken.None, TaskCreationOptions.DenyChildAttach, ts ?? TaskScheduler.Default);
+                }
+            }
+        }
+
+        public bool TrySetResult(T result)
+        {
+            if (Interlocked.CompareExchange(ref _state, (int)States.CompletionReserved, (int)States.Owned) == (int)States.Owned)
```

Comment:
I'm not entirely clear on why you have the CompletionReserved state.  What is this guarding against?  I suppose it prevents accidentally overwriting the result if this is called twice, but it seems like you're already in a bad place if TrySetResult is called twice.

Also, why return false here as opposed to throw?  What's the user of this supposed to do with a return value of false?

Reply:
> I'm not entirely clear on why you have the CompletionReserved state.

I had a reason, which now escapes me.  I'm going to leave it for now, as it's tied in to the whole state machine, but can subsequently look into removing it.

> What's the user of this supposed to do with a return value of false?

There can be multiple operations racing to complete one of these if cancellation is involved.  The synchronization and bool return is used in coordinating that.

Reply:
> I had a reason, which now escapes me.

Oh, right, it was also because of cancellation.  Cancellation races with another completion, the completion is about to store the value, cancellation comes along and completes the operation, the awaiter for that operation wakes up and retrieves the cancellation result, the object goes back to being pooled, another operation comes along and starts using it, and then that original operation writes its result over someone else's object.  Before we do any writing, we need to ensure we own the object.